add schema.org events to head

paaragon opened the pull request:
## Changes

Fixes #3252 

## Implementation details

Add an `application/ld+json` script in the head of `events/index.html` page o add the events information.

## How to read this PR

Run the page in dev mode and look for the script created within the HTML head tag. The correctness of the content can be tested with this page: https://search.google.com/test/rich-results?hl=es

y4izus reviewed (changes requested):
As mentioned on the issue, we are using the [useSchemaOrg](https://github.com/Qiskit/qiskit.org/pull/3374) composable, could you use it here, please?

paaragon commented:
With the last update, google finds 20 valid elements (the correct amount):

https://search.google.com/test/rich-results/result/r%2Fevents?id=z99Wo6f5cTegIPkGur1EGA&hl=es

![image](https://github.com/Qiskit/qiskit.org/assets/7396204/b04c7d87-eee3-4759-b595-fac432ae7571)

eddybrando commented on nuxt.config.ts:
```diff
@@ -1,10 +1,12 @@
 
 const { AIRTABLE_API_KEY, GENERATE_CONTENT, NODE_ENV, SITE_URL } = process.env;
 const IS_PRODUCTION = NODE_ENV === "production";
+const siteUrl = SITE_URL || "https://qiskit.org";
```
Do we need this to be variable?

I'm thinking that maybe we do because of our testing environment, but I'm not sure if it makes a difference or if we could just set it to "https://qiskit.org" always.

paaragon replied:
There was [an issue](https://github.com/Qiskit/qiskit.org/issues/3235) for change this from static to variable. Here, I just reorder it

eddybrando reviewed (changes requested):
Thanks a lot @paaragon, this is very good progress!

I added some specific comments, but also, generally, I believe that there is an issue with the current approach.

We shouldn't add the list of events to every page but only to the pages where they are listed. And also, for individual event pages, we should add the data for the individual event.

paaragon commented:
> Thanks a lot @paaragon, this is very good progress!
> 
> I added some specific comments, but also, generally, I believe that there is an issue with the current approach.
> 
> We shouldn't add the list of events to every page but only to the pages where they are listed. And also, for individual event pages, we should add the data for the individual event.

You are right, the events list should be referenced only on the event/index page (I have changed it).

I will also add the schema on `fall-fest`, `seminar-series` and `summer school (2022 and 2023)`

paaragon commented on composables/useQiskitSchemaOrg.ts:
```diff
@@ -0,0 +1,81 @@
+import { CommunityEvent } from "../types/events";
+import rawUpcomingEvents from "~/content/events/upcoming-community-events.json";
```
This file no longer exists

eddybrando commented on utils/event-schemaorg.ts:
```diff
@@ -0,0 +1,52 @@
+export interface EventSchemaOrg {
```
I believe that we're not importing this interface anywhere else, so we can remove the export:

```suggestion
interface EventSchemaOrg {
```

eddybrando reviewed (changes requested):
Thanks a lot @paaragon! It's looking very good 🙂

I left a couple of comments.

paaragon commented on utils/event-schemaorg.ts:
```diff
@@ -0,0 +1,52 @@
+      ? "OnlineEventAttendanceMode"
+      : "OfflineEventAttendanceMode";
+
+  const schemaEvent: any = {
```
Done!

paaragon commented on pages/events/index.vue:
```diff
@@ -282,6 +280,36 @@ const selectTab = (selectedTab: string) => {
+    const dateB = new Date(b.startDate);
+
+    return dateA > dateB ? 1 : dateA < dateB ? -1 : 0;
+  });
```
Done!

eddybrando reviewed (changes requested):
There are some issues after updating the branch. I'm not sure what they are related to, but might be due to the Nuxt update.

In any case, when looking at the dependencies, I realised that we're using the package `nuxt-schema-org`. According to them:

> The package `@vueuse/schema-org` has been superseded by [@unhead/schema-org](https://github.com/harlan-zw/unhead-schema-org) which offers the same simple API.

Source: https://nuxt.com/modules/schema-org#new-package

I think that as part of this PR we should replace the package. Maybe that would also bring us closer to the issue after Nuxt's update, which is that `useEventListSchemaOrg` is not found.

paaragon commented:
> There are some issues after updating the branch. I'm not sure what they are related to, but might be due to the Nuxt update.
> 
> In any case, when looking at the dependencies, I realised that we're using the package `nuxt-schema-org`. According to them:
> 
> > The package `@vueuse/schema-org` has been superseded by [@unhead/schema-org](https://github.com/harlan-zw/unhead-schema-org) which offers the same simple API.
> 
> Source: https://nuxt.com/modules/schema-org#new-package
> 
> I think that as part of this PR we should replace the package. Maybe that would also bring us closer to the issue after Nuxt's update, which is that `useEventListSchemaOrg` is not found.

The issue wasn't about the package but with some bug in the code. 

If you look at unhead's documentation on [how to integrate with nuxt](https://unhead-schema-org.harlanzw.com/integrations/nuxt/module), it says to use the nuxt-schema-org package

eddybrando commented:
> If you look at unhead's documentation on [how to integrate with nuxt](https://unhead-schema-org.harlanzw.com/integrations/nuxt/module), it says to use the nuxt-schema-org package

But on that page if you click on the GitHub icon on top, you land in the new repo https://github.com/harlan-zw/unhead-schema-org, which is `@unhead/schema-org`.

I believe they just haven't updated their documentation. But, to be fair, it's unclear either way.

I've opened an issue on their repo to ask about that: https://github.com/harlan-zw/unhead-schema-org/issues/24

eddybrando commented on components/EventsFallFest/EventsFallFestUniversityDirectory.vue:
```diff
@@ -43,6 +43,7 @@
 
 <script setup lang="ts">
 import { TextLink } from "~/types/links";
+import { sortEvents } from "~/utils/event-schemaorg";
```
Exports in the `utils` directory are globally available, so we don't need to import it.

With that in mind, I believe that it's a good idea to have very descriptive names for these functions and files.

Since this particular function sorts all kind of events, I think that it's indeed a good idea to call it `sortEvents`, but I also think that it should be in its own file (something like **utils/sortEvents.ts** and not included in a file related to the Schema.org stuff.

paaragon replied:
Thanks! I think is a good idea. Changes applied!

eddybrando commented on pages/events/index.vue:
```diff
@@ -124,6 +121,7 @@ import {
 
 import rawPastEvents from "~/content/events/past-community-events.json";
 import rawUpcomingEvents from "~/content/events/upcoming-community-events.json";
+import { sortEvents } from "~/utils/event-schemaorg";
```
Same comment as in the **components/EventsFallFest/EventsFallFestUniversityDirectory.vue** file:

> Exports in the `utils` directory are globally available, so we don't need to import it.
> 
> With that in mind, I believe that it's a good idea to have very descriptive names for these functions and files.
> 
> Since this particular function sorts all kind of events, I think that it's indeed a good idea to call it `sortEvents`, but I also think that it should be in its own file (something like **utils/sortEvents.ts** and not included in a file related to the Schema.org stuff.

eddybrando commented on utils/event-schemaorg.ts:
We introduce a file in kebab-case here. The other file in the **utils** folder is in camel case. Neither one is better than the other IMO and I'm fine with both ways. But we need to be consistent, so I think that we need to either rename this file to **eventSchemaorg.ts** or rename the other file to **url-has-protocol.ts**.

When I wrote the **urlHasProtocols.ts** file I had in mind that the spirit of the **utils** folder was to put individual reusable utils, being mostly files with 1 single exported function. If that's the case, then it would be nice to have the files have the same name as the exported function, so that they are easy to find.

paaragon replied:
I think we should use kebab-case as we are using in the hooks directory.

I agree with the only one exported function per file 😄

eddybrando reviewed (changes requested):
Thanks a lot @paaragon! I added a couple of comments 🙂

paaragon commented on layouts/default-max.vue:
Deleted!

eddybrando reviewed (approved):
Thanks a lot @paaragon! I just checked the rich results in the tool linked as well and it looks great.
